cloudpickle/cloudpickle_fast.py

```diff
@@ -478,6 +480,8 @@ def reducer_override(self, obj):
             return _class_reduce(obj)
         elif isinstance(obj, types.FunctionType):
             return self._function_reduce(obj)
+        elif isinstance(obj, typing.TypeVar) and not _is_importable_by_name(obj):
```

Comment:
`TypeVar` objects should be reduced using the `dispatch_table`, not the `reducer_override` hook, which is only used for objects such as functions and classes, for which the `dispatch_table` of a `Pickler` cannot be used.

Reply:
That was the original implementation, and I tried to fall back to `Pickler`'s implementation for global `TypeVar`s by doing this:

```
def _typevar_reduce(obj):
	if _is_importable_by_name(obj):
		return NotImplemented
	return (_make_typevar, _decompose_typevar(obj))
```

(like in `_function_reduce`, `_class_reduce`) but that fails with:

```
self = <cloudpickle.cloudpickle_fast.CloudPickler object at 0x000002B3BAEFEE80>, obj = ~T

    def dump(self, obj):
        try:
>               return Pickler.dump(self, obj)
E     _pickle.PicklingError: __reduce__ must return a string or tuple

cloudpickle\cloudpickle_fast.py:557: PicklingError
```

Here's a reproducing example:

```
from tests.mypkg import T
import io, pickle, cloudpickle
c = cloudpickle.CloudPickler(io.BytesIO())
_ = pickle.Pickler.dump(c, T)
```

I've looked into this for an hour so far, and haven't figured out why. Some findings:

- `T.__reduce__()` returns `'T'`, indicating it's a global
- this is the C code it's failing in: https://github.com/python/cpython/blob/3.8/Modules/_pickle.c#L4467; which means either `'T'` is failing the `PyUnicode_Check` check, or `reduce_value` is something else entirely
- forcing it to use the pure-Python pickler (`pickle._Pickler`) fails with a different error:

```
Traceback (most recent call last):
  File "foo.py", line 12, in <module>
    _ = Pickler.dump(c, T)
  File "D:\dev\python38\lib\pickle.py", line 485, in dump
    self.save(obj)
  File "D:\dev\python38\lib\pickle.py", line 558, in save
    f(self, obj)  # Call unbound method with explicit self
TypeError: _typevar_reduce() takes 1 positional argument but 2 were given
```

---

Because of this, the only way I could get it to fall-back correctly was by handling `TypeVar` in `reducer_override`.

Reply:
Actually, you should never return `NotImplemented` when trying to reduce a `TypeVar`: If you encounter an importable `TypeVar`, you should make `_reduce_typevar  return something like (`getattr, module_and_name`), like in `cloudpickle.py`.
Triggering fallbacks using `NotImplemented` only exist for `reducer_override`, which has special semantics because it gets called very early in the pickling process.

Reply:
That's a shame, so there's no way to fall back to `pickle`'s implementation from inside a reducer?

Reply:
As far as I know the Pickle API does not allow for that.

That would be a nice addition to the standard library though, I think. But maybe not trivial to implement efficiently? I don't know.

Reply:
Bear in mind that `TypeVar` objects are standard instances - like all instances deriving from the `object` class, they have a default `__reduce__` method. So falling back in `dispatch` can simply done using `return obj.__reduce__()`.